src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,846 @@
+                }
+
+                private PooledListForMatchResult(List<MatchResult<VSCompletionItem>> list)
+                {
+                    _list = list;
+                }
+
+                public void Dispose()
+                {
+                    if (_list is not null)
+                    {
+                        // Don't call ClearAndFree, which resets the capacity to a default value.
+                        _list.Clear();
+                        s_listOfMatchResultPool.Free(_list);
+                        _list = null;
+                    }
+                }
+            }
+
+            private sealed class FilterStateHelper
+            {
+                private readonly ImmutableArray<CompletionFilterWithState> _nonExpanderFilterStates;
+                private readonly ImmutableArray<CompletionFilter> _selectedNonExpanderFilters;
+                private readonly bool _needToFilter;
+                private readonly ImmutableArray<CompletionFilter> _unselectedExpanders;
+                private readonly bool _needToFilterExpanded;
+
+                // For telemetry
+                private readonly object _targetTypeCompletionFilterChosenMarker = new();
+
+                public FilterStateHelper(ImmutableArray<CompletionFilterWithState> filtersWithState)
+                {
+                    // We need to filter if 
+                    // 1. a non-empty strict subset of filters are selected
+                    // 2. a non-empty set of expanders are unselected
+                    _nonExpanderFilterStates = filtersWithState.WhereAsArray(f => f.Filter is not CompletionExpander);
+
+                    _selectedNonExpanderFilters = _nonExpanderFilterStates.SelectAsArray(f => f.IsSelected, f => f.Filter);
+                    _needToFilter = _selectedNonExpanderFilters.Length > 0 && _selectedNonExpanderFilters.Length < _nonExpanderFilterStates.Length;
+
+                    _unselectedExpanders = filtersWithState.SelectAsArray(f => !f.IsSelected && f.Filter is CompletionExpander, f => f.Filter);
+                    _needToFilterExpanded = _unselectedExpanders.Length > 0;
+                }
+
+                public bool ShouldBeFilteredOut(VSCompletionItem item)
+                    => ShouldBeFilteredOutOfCompletionList(item) || ShouldBeFilteredOutOfExpandedCompletionList(item);
+
+                private bool ShouldBeFilteredOutOfCompletionList(VSCompletionItem item)
+                    => _needToFilter && !item.Filters.Any(filter => _selectedNonExpanderFilters.Contains(filter));
+
+                private bool ShouldBeFilteredOutOfExpandedCompletionList(VSCompletionItem item)
+                {
+                    if (!_needToFilterExpanded)
+                        return false;
+
+                    var associatedWithUnselectedExpander = false;
+                    foreach (var itemFilter in item.Filters)
+                    {
+                        if (itemFilter is CompletionExpander)
```

Comment:
what is a completion-expander?

Reply:
It's this plus button at the beginning of the filter list

![image](https://user-images.githubusercontent.com/788783/144918975-88d6593c-d066-4994-a21c-dcccddeb8dd3.png)